pytorch_lightning/plugins/training_type/deepspeed.py

```diff
@@ -312,7 +313,15 @@ def _initialize_deepspeed_train(self, model):
     @contextlib.contextmanager
     def model_sharded_context(self) -> Generator[None, None, None]:
         if self.zero_stage_3:
-            model_parallel_context = deepspeed.zero.Init(remote_device="cpu", pin_memory=True)
+            kwargs = {
+                "remote_device": "cpu",
+                "pin_memory": True,
+            }
+            # from DeepSpeed 0.4.0, weights need to be properly casted before calling `deepspeed.initialize`.
+            if "dtype" in inspect.signature(deepspeed.zero.Init).parameters.keys():
+                precision: int = self.lightning_module.trainer.accelerator.precision
+                kwargs["dtype"] = torch.float16 if precision == 16 else torch.float32
+            model_parallel_context = deepspeed.zero.Init(**kwargs)
```

Comment:
Thanks for this! I'd like to properly address this in #7234 so do you mind if I bring this fix into that PR?

Reply:
@SeanNaren Yes, go for it. Merge this PR in yours.